### PR TITLE
feat: Add linter for AIP-192 deprecated service and methods

### PR DESCRIPTION
**Pull request by miraleung**
Corresponds to [this AIP PR](https://github.com/aip-dev/google.aip.dev/pull/736).

Note to future me: This was my first Golang PR.

### CODE REVIEW

File: docs/rules/0192/deprecated-method-comment.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+--------------------------------------------------------------------------------
+
+rule: aip: 192 name: [core, '0192', deprecated-method-comment] summary:
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:3]: This section isn't formatted the same way as the other rule docs, e.g.

https://raw.githubusercontent.com/googleapis/api-linter/main/docs/rules/0192/only-leading-comments.md

File: docs/rules/0192/deprecated-method-comment.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Deprecated methods must have a corresponding comment. permalink:
+/192/deprecated-method-comment redirect_from:
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:5]: This should redirect from `/0192/deprecated-method-comment`.

File: docs/rules/0192/deprecated-method-comment.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+## - /0192/deprecated-method-comment
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:7]:
```suggestion
```

File: docs/rules/0192/deprecated-method-comment.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+# Omnipresent comments
+
+This rule enforces that every RPC marked with the protobuf `deprecated`option
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:11]:
```suggestion
This rule enforces that every RPC marked with the protobuf `deprecated` option
```

File: rules/aip0192/deprecated.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+var deprecatedMethodComment = &lint.MethodRule{
+	Name:   lint.NewRuleName(192, "deprecated-method-comment"),
+	OnlyIf: isDeprecatedMethod,
+	LintMethod: func(m *desc.MethodDescriptor) []lint.Problem {
+		comment := separateInternalComments(m.GetSourceInfo().GetLeadingComments())
+		if len(comment.External) > 0 && strings.HasPrefix(comment.External[0], "Deprecated:") {
+			return nil
+		}
+		return []lint.Problem{{
+			Message:    "Deprecated methods must have \"Deprecated: <reason>\" as the first line in the method comment.",
```
Comment (apasel422) [inline rules/aip0192/deprecated.go:33]:
```suggestion
			Message:    `Deprecated methods must have "Deprecated: <reason>" as the first line in the method comment.`,
```

File: rules/aip0192/deprecated_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            option deprecated = true;
+          }
+        }
+        message GetBookRequest {}
+        message Book {}
+      `, test)
+
+			problems := deprecatedMethodComment.Lint(file)
+			if diff := test.problems.SetDescriptor(file.GetServices()[0].GetMethods()[0]).Diff(problems); diff != "" {
+				t.Errorf(diff)
```
Comment (apasel422) [inline rules/aip0192/deprecated_test.go:50]:
```suggestion
				t.Error(diff)
```

File: rules/aip0192/deprecated_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+// These are split up since templating doesn't play nicely with inserting protobuf options.
+func TestDeprecatedMethod(t *testing.T) {
+	tests := []struct {
+		testName      string
+		MethodComment string
+		problems      testutils.Problems
+	}{
+		{"ValidMethodDeprecated", "// Deprecated: Don't use this.\n// Method comment.", nil},
+		{"InvalidMethodDeprecated", "// Method comment.", testutils.Problems{{Message: "Deprecated methods"}}},
```
Comment (apasel422) [inline rules/aip0192/deprecated_test.go:31]: We should add a case to ensure that the lint doesn't return a problem if the method is *not* deprecated.

File: docs/rules/0192/deprecated-method-comment.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+has `"Deprecated: <reason>"` as the first line in the public leading comment, as
+mandated in [AIP-192][].
+
+## Details
+
+This rule looks at each method descriptor in each proto file, and complains if
+the protobuf `deprecated` option is set, but the first line of the public
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:18]:
```suggestion
the protobuf `deprecated` option is set to `true`, but the first line of the public
```

File: docs/rules/0192/deprecated-method-comment.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+If you need to violate this rule, use a leading comment above the descriptor.
+Remember to also include an [aip.dev/not-precedent][] comment explaining why.
+
+```proto
+// A representation of a book.
+message Book {
+  // (-- api-linter: core::0192::deprecated-method-comment=disabled
+  //     aip.dev/not-precedent: We need to do this because reasons. --)
+  string name = 1;
+}
```
Comment (apasel422) [inline docs/rules/0192/deprecated-method-comment.md:61]: This code should be based on the incorrect example above.

File: rules/aip0192/deprecated.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+			Descriptor: m,
+		}}
+	},
+}
+
+var deprecatedServiceComment = &lint.ServiceRule{
```
Comment (apasel422) [inline rules/aip0192/deprecated.go:39]: Do we actually need separate rules for services and methods? It seems like these can be combined using a `lint.DescriptorRule` that runs only if the `desc.Descriptor` is a `*desc.ServiceDescriptor` or `*desc.MethodDescriptor`. This would also make it easier to apply to messages and fields, if the AIP is ever amended with similar guidance for those.

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: rules/aip0192/aip0192.go [hunk 11]
```diff
@@ -35,6 +38,22 @@ func AddRules(r lint.RuleRegistry) error {
 	)
 }
 
+// Returns true if this is a deprecated method, false otherwise.
+func isDeprecatedMethod(m *desc.MethodDescriptor) bool {
+	if m.GetMethodOptions() == nil {
+		return false
+	}
+	return *m.GetMethodOptions().Deprecated
```
Comment (noahdietz) [inline rules/aip0192/aip0192.go:46]: IDK off the top of my head, but is there a `GetDeprecated` accessor? If so, please use that instead of direct field access.

File: rules/aip0192/aip0192.go [hunk 12]
```diff
@@ -35,6 +38,22 @@ func AddRules(r lint.RuleRegistry) error {
...
+}
+
+// Returns true if this is a deprecated service, false otherwise.
+func isDeprecatedService(s *desc.ServiceDescriptor) bool {
+	if s.GetServiceOptions() == nil {
+		return false
+	}
+	return *s.GetServiceOptions().Deprecated
```
Comment (noahdietz) [inline rules/aip0192/aip0192.go:54]: Same thing here WRT use of accessor method.

File: .gitignore [hunk 13]
```diff
@@ -5,3 +5,7 @@
 docs/default.profraw
 docs/_site
 docs/.jekyll-metadata
+
+# Vim
```
Comment (noahdietz) [inline .gitignore:9]: This can be reverted I merged your separate PR.

File: rules/aip0192/aip0192.go (hunk 11 above)
Comment (apasel422) [reply rules/aip0192/aip0192.go:46]: Yeah, I think we just want:

```go
return m.GetMethodOptions().GetDeprecated()
```

File: docs/rules/0192/deprecated-method-comment.md (hunk 1 above)
Comment (miraleung) [reply docs/rules/0192/deprecated-method-comment.md:3]: Done.

File: docs/rules/0192/deprecated-method-comment.md (hunk 2 above)
Comment (miraleung) [reply docs/rules/0192/deprecated-method-comment.md:5]: Done.

File: docs/rules/0192/deprecated-method-comment.md (hunk 4 above)
Comment (miraleung) [reply docs/rules/0192/deprecated-method-comment.md:11]: Done.

File: rules/aip0192/deprecated.go (hunk 5 above)
Comment (miraleung) [reply rules/aip0192/deprecated.go:33]: Done.

File: rules/aip0192/deprecated_test.go (hunk 6 above)
Comment (miraleung) [reply rules/aip0192/deprecated_test.go:50]: Done.

File: docs/rules/0192/deprecated-method-comment.md (hunk 8 above)
Comment (miraleung) [reply docs/rules/0192/deprecated-method-comment.md:18]: Done.

File: docs/rules/0192/deprecated-method-comment.md (hunk 9 above)
Comment (miraleung) [reply docs/rules/0192/deprecated-method-comment.md:61]: Done.

File: rules/aip0192/deprecated.go (hunk 10 above)
Comment (miraleung) [reply rules/aip0192/deprecated.go:39]: Done.

File: rules/aip0192/aip0192.go (hunk 11 above)
Comment (miraleung) [reply rules/aip0192/aip0192.go:46]: There doesn't seem to be one.

File: rules/aip0192/aip0192.go (hunk 12 above)
Comment (miraleung) [reply rules/aip0192/aip0192.go:54]: Acknowledged.

File: .gitignore (hunk 13 above)
Comment (miraleung) [reply .gitignore:9]: Done.

File: rules/aip0192/aip0192.go [hunk 14]
```diff
@@ -35,6 +37,26 @@ func AddRules(r lint.RuleRegistry) error {
...
 
+// Returns true if this is a deprecated method, false otherwise.
+func isDeprecated(d desc.Descriptor) bool {
+	switch d.(type) {
+	case *desc.MethodDescriptor:
+		m := d.(*desc.MethodDescriptor)
+		if m.GetMethodOptions() == nil {
+			return false
+		}
+		return *m.GetMethodOptions().Deprecated
```
Comment (apasel422) [inline rules/aip0192/aip0192.go:48]:
```suggestion
		return d.GetMethodOptions().GetDeprecated()
```

File: rules/aip0192/aip0192.go [hunk 15]
```diff
@@ -35,6 +37,26 @@ func AddRules(r lint.RuleRegistry) error {
...
+	case *desc.ServiceDescriptor:
+		s := d.(*desc.ServiceDescriptor)
+		if s.GetServiceOptions() == nil {
+			return false
+		}
+		return *s.GetServiceOptions().Deprecated
```
Comment (apasel422) [inline rules/aip0192/aip0192.go:54]:
```suggestion
		return d.GetServiceOptions().GetDeprecated()
```

File: rules/aip0192/aip0192.go [hunk 16]
```diff
@@ -35,6 +37,26 @@ func AddRules(r lint.RuleRegistry) error {
 	)
 }
 
+// Returns true if this is a deprecated method, false otherwise.
```
Comment (apasel422) [inline rules/aip0192/aip0192.go:40]:
```suggestion
// Returns true if this is a deprecated method or service, false otherwise.
```

File: rules/aip0192/deprecated_comment.go [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+var deprecatedComment = &lint.DescriptorRule{
+	Name:   lint.NewRuleName(192, "deprecated-comment"),
+	OnlyIf: isDeprecated,
+	LintDescriptor: func(d desc.Descriptor) []lint.Problem {
+		comment := separateInternalComments(d.GetSourceInfo().GetLeadingComments())
+		if len(comment.External) > 0 && strings.HasPrefix(comment.External[0], "Deprecated:") {
+			return nil
+		}
+		return []lint.Problem{{
+			Message:    `Use "Deprecated: <reason>" as the first line in the  comment.`,
```
Comment (apasel422) [inline rules/aip0192/deprecated_comment.go:33]:
```suggestion
			Message:    `Use "Deprecated: <reason>" as the first line in the comment.`,
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by noahdietz — COMMENTED** [review]
LGTM with @apasel422's approval

File: rules/aip0192/aip0192.go (hunk 15 above)
Comment (apasel422) [inline rules/aip0192/aip0192.go:54]:
```suggestion
	switch d := d.(type) {
	case *desc.MethodDescriptor:
		return d.GetMethodOptions().GetDeprecated()
	case *desc.ServiceDescriptor:
		return d.GetServiceOptions().GetDeprecated()
```

File: rules/aip0192/deprecated_comment.go (hunk 17 above)
Comment (miraleung) [reply rules/aip0192/deprecated_comment.go]: Done.

File: rules/aip0192/aip0192.go (hunk 15 above)
Comment (miraleung) [reply rules/aip0192/aip0192.go]: This fails with the error message below. The `Options() == nil` check is also necessary so that the valid case (without `deprecated = true`) won't result in a null pointer deference error.

Error message: `d.GetMethodOptions undefined (type desc.Descriptor has no field or method GetMethodOptions)`

File: rules/aip0192/deprecated_test.go (hunk 7 above)
Comment (miraleung) [reply rules/aip0192/deprecated_test.go]: Done.

File: rules/aip0192/aip0192.go (hunk 15 above)
Comment (apasel422) [reply rules/aip0192/aip0192.go]: I updated my suggestion to fix the error. But the options check isn't necessary because the generated Go code already handles the `nil` case.

File: rules/aip0192/aip0192.go [hunk 18]
```diff
@@ -35,6 +37,26 @@ func AddRules(r lint.RuleRegistry) error {
...
+		if m.GetMethodOptions() == nil {
+			return false
+		}
+		return m.GetMethodOptions().GetDeprecated()
+	case *desc.ServiceDescriptor:
+		s := d.(*desc.ServiceDescriptor)
+		if s.GetServiceOptions() == nil {
+			return false
+		}
+		return s.GetServiceOptions().GetDeprecated()
```
Comment (apasel422) [inline rules/aip0192/aip0192.go:54]:
```suggestion
	switch d := d.(type) {
	case *desc.MethodDescriptor:
		return d.GetMethodOptions().GetDeprecated()
	case *desc.ServiceDescriptor:
		return d.GetServiceOptions().GetDeprecated()
```

File: rules/aip0192/aip0192.go (hunk 15 above)
Comment (miraleung) [reply rules/aip0192/aip0192.go]: Thanks!

**Review by apasel422 — APPROVED** [review]

**Comment by lukesneeringer** [comment]
I am good with this once my question in https://github.com/aip-dev/google.aip.dev/pull/736 is decided. (If we decide to do "last paragraph", I expect it will require code updates here.)

**Comment by lukesneeringer** [comment]
aip-dev/google.aip.dev#736 is decided in favor of this implementation. This can go in as soon as that PR does.
